core/src/main/java/jenkins/model/Nodes.java

```diff
@@ -139,6 +140,7 @@ public void run() {
             });
             // TODO there is a theoretical race whereby the node instance is updated/removed after lock release
             persistNode(node);
+            NodeListener.fireOnCreated(node);
```

Comment:
Something nice to do for all listeners, but here we are adding a new listener API so no need to worry about breaking backward compatibility

Also in this case we have some concurrency concerns due to the lock proximity and the fact that we steal the calling thread for notification.

We would need to figure out what kind of event delivery model to apply. IMHO at the very least each listener should be guaranteed in-order serialised delivery of events. We could do that by having a single event notification thread or by having a pool but picking the listener's thread in the pool based on a hash